modules/stochastic_tools/src/samplers/AIS.C

```diff
@@ -0,0 +1,177 @@
+{
+  // Filling the `distributions` vector with the user-provided distributions.
+  for (const DistributionName & name : _distribution_names)
+    _distributions.push_back(&getDistributionByName(name));
+
+  /* Adaptive Importance Sampling (AIS) relies on a Markov Chain Monte Carlo (MCMC) algorithm.
+     As such, in MOOSE, any use of MCMC algorithms requires that the `num_steps` parameter in the
+     main App's executioner would control the total number of samples. Therefore, the `num_rows`
+     parameter typically used by exisiting non-MCMC samplers to set the total number of samples
+     has no use here and is fixed to 1.*/
+  int num_rows = 1;
+  setNumberOfRows(num_rows);
+
+  // Setting the number of columns in the sampler matrix (equal to the number of distributions).
+  setNumberOfCols(_distributions.size());
+
+  /* `inputs_sto` is a member variable that aids in forming the importance distribution.
+     One dimension of this variable is equal to the number of distributions. The other dimension
+     of the variable, at the last step, is equal to the number of samples the user desires.*/
+  _inputs_sto.resize(_distributions.size());
+
+  // Mapping all the input distributions to a standard normal space
+  for (unsigned int i = 0; i < _distributions.size(); ++i)
+    _inputs_sto[i].push_back(Normal::quantile(_distributions[i]->cdf(_initial_values[i]), 0, 1));
+
+  /* `prev_value` is a member variable for tracking the previously accepted samples in the
+     MCMC algorithm and proposing the next sample.*/
+  _prev_value.resize(_distributions.size());
+
+  // `check_step` is a member variable for ensuring that the MCMC algorithm proceeds in a sequential
+  // fashion.
+  _check_step = 0;
+
+  // Storage for means of input values for proposing the next sample
+  _mean_sto.resize(_distributions.size());
+
+  // Storage for standard deviations of input values for proposing the next sample
+  _std_sto.resize(_distributions.size());
+
+  setNumberOfRandomSeeds(_num_random_seeds);
+}
+
+Real
+AIS::computeSample(dof_id_type /*row_index*/, dof_id_type col_index)
+{
+  TIME_SECTION(_perf_compute_sample);
+
+  if (_step <= _num_samples_train)
+  {
+    /* This is the importance distribution training step. Markov Chains are set up
+       to sample from the importance region or the failure region using the Metropolis
+       algorithm. Given that the previous sample resulted in a model failure, the next
+       sample is proposed such that it is very likely to result in a model failure as well.
+       The `initial_values` and `proposal_std` parameters provided by the user affects the
+       formation of the importance distribution. */
+    if (_step > 1 && col_index == 0 && _check_step != _step)
+    {
+      for (dof_id_type j = 0; j < _distributions.size(); ++j)
+        _prev_value[j] = Normal::quantile(
+            _distributions[j]->cdf(getReporterValueByName<Real>(_inputs_names[j])), 0, 1);
```

Comment:
Instead of getting the reporter value every time, you can define a refence for it in the contructor. Like this:
In the header file:
```
const std::vector<Real> & _inputs;
```
In the constructor:
```
_inputs(getReporterValue<std::vector<Real>>("inputs");
```
This is if you change the value in the reporter to a vector like I suggested before.